files/nrpe/check_windows_instances.sh

```diff
@@ -0,0 +1,98 @@
+#!/bin/bash
+# Checks that instances running Windows are actually running in their designated aggregates
+
+# Written by Matteo Pozza @ CSC 2021
+
+
+## Nagios return codes
+WARNING=1
+#CRITICAL=2
+UNKNOWN=3
+OK=0
+
+# BUSTED will contain the list of instances that have been spotted outside their designated aggregate
+BUSTED=""
+# NL is just an easy variable to add a carriage return to a string
+NL=$'\n'
+
+usage ()
+{
+  echo "Usage: $0 [OPTIONS]"
+  echo " -h                 Get help"
+  echo " -u <username>      OS_USERNAME"
+  echo " -p <password>      OS_PASSWORD"
+  echo " -a <url>           OS_AUTH_URL"
+  echo " -t <project_name>  OS_TENANT_NAME"
+  echo " -e <domain_name>   OS_USER_DOMAIN_NAME and OS_PROJECT_DOMAIN_NAME"
+}
+
+# if there are no args, then give help and exit
+if (($# == 0)); then
+  usage
+  exit $UNKNOWN 
+fi
+
+# parse the args
+while getopts 'hu:p:a:t:e:' OPTION
+do
+    case $OPTION in
+        h)
+            usage
+            exit $UNKNOWN
+            ;;
+        u)
+            export OS_USERNAME=$OPTARG
+            ;;
+        p)
+            export OS_PASSWORD=$OPTARG
+            ;;
+        a)
+            export OS_AUTH_URL=$OPTARG
+            ;;
+        t)
+            export OS_TENANT_NAME=$OPTARG
+            ;;
+        e)
+            export OS_USER_DOMAIN_NAME=$OPTARG
+            export OS_PROJECT_DOMAIN_NAME=$OPTARG
+            ;;
+	    *)
+            usage
+            exit $UNKNOWN
+            ;;
+    esac
+done
+shift "$((OPTIND-1))"
+
+export OS_IDENTITY_API_VERSION=3
+
+# get the id of the aggregates whose name contains "win"
+for AGG_ID in $(openstack aggregate list -c ID -c Name -f value | grep -i win | awk '{print $1}'); do
+    # prepare a variable of the form "host1|host2|host3", where host1, host2, host3 are the hosts of the aggregate
+    NODES_LIST=""
+    for NODE in $(openstack aggregate show "$AGG_ID" -c hosts -f value | sed -e "s/\[u'//g" -e "s/ u'/ /g" -e "s/\[//g" -e "s/\]//g" -e "s/,//g"  -e "s/'//g"); do
```

Comment:
In the spirit of "simple is beautiful",

1) While I like a good sed spell as much as the next guy, if changing to `-f json`, this might be much more readable for others.

2) Optionally expanding on 1), **if** it's reasonable to presume the host has `jq` installed, this could be further simplified with `-f json|jq --raw-output '.hosts | .[]'`

3) You would also be able to lose the awk in AGG_ID collection if you'd get the names instead of ID's. But admittedly, only the IDs are unique whereas names are not, so this would not be as robust.

Reply:
Thanks for the comments.
About 1 and 2: I propose you the following
```
-f yaml | tail -n+2 | sed "s/- //g"
```
1) it is way less convoluted than the previous sed expression
2) similarly to `-f json`, `-f yaml` provides the same output regardless of python2 or python3 client (no 'u' chars in front of utf-8 strings in case of python2)
3) it does not require to install additional packages

About 3: I think we should stick with using ids whenever possible. Also, the awk expression is not very convoluted. I can put a comment above explaining what it does, if you think this would help.

Reply:
Sounds great! :+1: 

No need to comment item 3, it's fine, ID's are king